### PR TITLE
Add reduction of rules by coverage

### LINKED ISSUE #277: Add LF filtering functionality
Add function to filter LF by coverage and adjust the z and t matrices accordingly.

### PR DESCRIPTION
**Pull request by marina-sp**
closes #277

### CODE REVIEW

**Comment by marina-sp** [comment]
This can be used for making the dataset smaller wrt the number of rules. E.g. I intend to use it for Tacred where there are too many rules for Snorkel to handle. A (hopefully) detailed description of the reduction process is provided in the code comments.

I am not sure about the placement of this function, since it should be mostly used for data preprocessing prior to passing it to a Trainer. Then it would make sense to place it in the experiments.utils or similar. However, we could include this reduction into the Snorkel Trainer itself to ensure the data is reasonably sized before calling the LabelModel. Then it is appropriate to leave this functionality in transformations (as it is now). I would appreciate your thoughts on this :)

File: knodle/transformation/rule_reduction.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    "returning the original LF matches.")
+        return {"rule_matches_z": rule_matches_z, "mapping_rule_class_t": mapping_rule_class_t}
+
+    coverage_per_rule = rule_matches_z.sum(0) / rule_matches_z.shape[0]
+
+    # create mask to indicate which rules will be kept unchanged
+    lf_kept_mask = np.zeros(mapping_rule_class_t.shape[0], dtype=np.bool)
+
+    # take top N rules
+    if max_rules is not None:
```
Comment (agsedova) [inline knodle/transformation/rule_reduction.py:44]: to be sure: is it a valid use case when user wants both filtering by max rules and coverage?

File: knodle/transformation/rule_reduction.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    if rule_matches_rest is not None:
+        rule_matches_dict.update(rule_matches_rest)
+
+    if drop_rules:
+        return _reduce_by_drop(rule_matches_dict, mapping_rule_class_t, lf_kept_mask)
+    else:
+        return _reduce_by_merge(rule_matches_dict, mapping_rule_class_t, lf_kept_mask)
+
+
+def _reduce_by_drop(rule_matches_dict, mapping_rule_class_t, lf_kept_mask):
```
Comment (agsedova) [inline knodle/transformation/rule_reduction.py:63]: typing and output please :) a couple of comment words would also be good

File: knodle/transformation/rule_reduction.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    output_dict = {}
+    for split, match_matrix in rule_matches_dict.items():
+        output_dict[split] = match_matrix[:, lf_kept_mask]
+    output_dict["mapping_rule_class_t"] = mapping_rule_class_t[lf_kept_mask, :]
+    return output_dict
+
+
+def _reduce_by_merge(rule_matches_dict, mapping_rule_class_t, lf_kept_mask):
```
Comment (agsedova) [inline knodle/transformation/rule_reduction.py:71]: typing and output

File: knodle/transformation/rule_reduction.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    output_dict["mapping_rule_class_t"] = np.vstack([output_dict["mapping_rule_class_t"], merged_mapping])
+
+    return output_dict
+
+
+def _get_lf_by_label_iterator(
+    mapping_rule_class_t: np.ndarray
+) -> Iterable[np.ndarray]:
+    """
+    Get an iterator yielding an rule ids mask for every label, indicating which rules belong to this label.
```
Comment (agsedova) [inline knodle/transformation/rule_reduction.py:101]: an -> a

Comment (agsedova) [reply knodle/transformation/rule_reduction.py:101]: btw there is a function `get_rules_sample_ids` in knodle/trainer/wscrossweigh - I guess it is done pretty much the same, but not exactly.

**Review by agsedova — CHANGES_REQUESTED** [review]

**Comment by agsedova** [comment]
> I am not sure about the placement of this function, since it should be mostly used for data preprocessing prior to passing it to a Trainer. Then it would make sense to place it in the experiments.utils or similar. However, we could include this reduction into the Snorkel Trainer itself to ensure the data is reasonably sized before calling the LabelModel. Then it is appropriate to leave this functionality in transformations (as it is now). I would appreciate your thoughts on this :)

I don't think it should be a part of `experiments.utils` or whatever, but a part of the knodle core. Totally fine with where it is located now, in`transformations`.

**Review by agsedova — APPROVED** [review]

**Pull request merged**